src/System.Collections/tests/Generic/Dictionary/Dictionary.Generic.Tests.netcoreapp.cs

```diff
@@ -161,5 +162,181 @@ public void EnsureCapacity_SetsToNextPrimeNumber()
         }
 
         #endregion
+
+        #region TrimExcess
+
+        [Fact]
+        public void TrimExcess_NegativeCapacity_Throw()
+        {
+            var dictionary = new Dictionary<int, int>();
+            Assert.Throws<ArgumentOutOfRangeException>(() => dictionary.TrimExcess(-1));
+        }
+
+        [Fact]
+        public void TrimExcess_CapacitySmallerThanCount_Throws()
+        {
+            var dictionary = new Dictionary<int, int>();
+            dictionary.Add(1, 1);
+            Assert.Throws<ArgumentOutOfRangeException>(() => dictionary.TrimExcess(0));
+
+            dictionary = new Dictionary<int, int>(20);
+            dictionary.Add(1, 1);
+            Assert.Throws<ArgumentOutOfRangeException>(() => dictionary.TrimExcess(0));
+        }
+
+        [Fact]
+        public void TrimExcess_LargeCapacity_TrimsSuccessfully()
+        {
+            var dictionary = new Dictionary<int, int>(20);
+            dictionary.TrimExcess(7);
+            Assert.Equal(7, dictionary.EnsureCapacity(0));
+        }
+
+        [Fact]
+        public void TrimExcess_NoArguments_TrimsToAtLeastCount()
+        {
+            var dictionary = new Dictionary<int, int>(20);
+            for (int i = 0; i < 7; i++)
+            {
+                dictionary.Add(i, i);
+            }
+            dictionary.TrimExcess(7);
+            Assert.Equal(7, dictionary.EnsureCapacity(0));
+
+            dictionary = new Dictionary<int, int>(20);
+            for (int i = 0; i < 8; i++)
+            {
+                dictionary.Add(i, i);
+            }
+            dictionary.TrimExcess(8);
+            Assert.Equal(11, dictionary.EnsureCapacity(0));
+        }
+
+        [Fact]
+        public void TrimExcess_TrimToLargerThanExistingCapacity_DoesNothing()
+        {
+            var dictionary = new Dictionary<int, int>();
+            int capacity = dictionary.EnsureCapacity(0);
+            dictionary.TrimExcess(20);
+            Assert.Equal(capacity, dictionary.EnsureCapacity(0));
+
+            dictionary = new Dictionary<int, int>(10);
+            capacity = dictionary.EnsureCapacity(0);
+            dictionary.TrimExcess(20);
+            Assert.Equal(capacity, dictionary.EnsureCapacity(0));
+        }
+
+        [Fact]
+        public void TrimExcess_EmptyDictionaryWithDefaultCapacity_CapacityRemainsAsMinPossible()
+        {
+            var dictionary = new Dictionary<int, int>();
+            Assert.Equal(3, dictionary.EnsureCapacity(0));
```

Comment:
Why does EnsureCapacity(0) on an empty dictionary force it to increase its array size?

Reply:
The Dictionary capacity size is only set to prime numbers and the smallest valid prime number to set to is 3. (refer to here)[https://github.com/dotnet/coreclr/blob/0fc8b13bfbf14a51964b95e29a67cbf53f8609c3/src/mscorlib/shared/System/Collections/HashHelpers.cs#L30]

When EnsureCapacity gets called, in case the Dictionary is not initialized yet, then the entries and buckets arrays will get set to the first prime number larger than the argument provided to EnsureCapacity.

Reply:
> When EnsureCapacity gets called, in case the Dictionary is not initialized yet, then the entries and buckets arrays will get set to the first prime number larger than the argument provided to EnsureCapacity.

Yes, but why when the argument is 0?  Is it just that we don't expect this to be a common case and thus we're not interested in optimizing it?  Logically it seems strange to me that if the dictionary's current capacity is 0 and we call EnsureCapacity(0) that it would then choose to allocate an array.

Reply:
I think it is fair to say that EnsureCapacity should return the current capacity and only if the argument is larger than existing capacity then is should update capacity.

Based on the above logic then I think you are right: in the case when dictionary is not yet initialized, EnsureCapacity(0) better just return zero as capacity without initializing the Dictionary.

Reply:
I opened up this coreclr PR to make this modification on EnsureCapacity: https://github.com/dotnet/coreclr/pull/16076